pkg/core/resources/apis/workload/generate/generator.go

```diff
@@ -0,0 +1,181 @@
+	mesh_proto "github.com/kumahq/kuma/v2/api/mesh/v1alpha1"
+	core_mesh "github.com/kumahq/kuma/v2/pkg/core/resources/apis/mesh"
+	workload_api "github.com/kumahq/kuma/v2/pkg/core/resources/apis/workload/api/v1alpha1"
+	"github.com/kumahq/kuma/v2/pkg/core/resources/manager"
+	"github.com/kumahq/kuma/v2/pkg/core/resources/model"
+	"github.com/kumahq/kuma/v2/pkg/core/resources/store"
+	"github.com/kumahq/kuma/v2/pkg/core/runtime/component"
+	"github.com/kumahq/kuma/v2/pkg/core/user"
+	core_metrics "github.com/kumahq/kuma/v2/pkg/metrics"
+	"github.com/kumahq/kuma/v2/pkg/plugins/runtime/k8s/metadata"
+	mesh_cache "github.com/kumahq/kuma/v2/pkg/xds/cache/mesh"
+	xds_context "github.com/kumahq/kuma/v2/pkg/xds/context"
+)
+
+const (
+	managedByValue = "workload-generator"
+)
+
+// Generator generates Workload objects from Dataplane resources created on
+// Universal.
+type Generator struct {
+	logger           logr.Logger
+	generateInterval time.Duration
+	metric           prometheus.Summary
+	resManager       manager.ResourceManager
+	meshCache        *mesh_cache.Cache
+	zone             string
+}
+
+var _ component.Component = &Generator{}
+
+func New(
+	logger logr.Logger,
+	generateInterval time.Duration,
+	metrics core_metrics.Metrics,
+	resManager manager.ResourceManager,
+	meshCache *mesh_cache.Cache,
+	zone string,
+) (*Generator, error) {
+	metric := prometheus.NewSummary(prometheus.SummaryOpts{
+		Name:       "component_workload_generator",
+		Help:       "Summary of Workload generation duration",
+		Objectives: core_metrics.DefaultObjectives,
+	})
+	if err := metrics.Register(metric); err != nil {
+		return nil, err
+	}
+	return &Generator{
+		logger:           logger,
+		generateInterval: generateInterval,
+		metric:           metric,
+		resManager:       resManager,
+		meshCache:        meshCache,
+		zone:             zone,
+	}, nil
+}
+
+// workloadForDataplane extracts and validates the workload identifier from a dataplane
+func (g *Generator) workloadForDataplane(dataplane *core_mesh.DataplaneResource) (string, bool) {
+	workloadName, ok := dataplane.GetMeta().GetLabels()[metadata.KumaWorkload]
+	if !ok || workloadName == "" {
+		return "", false
+	}
+	allErrs := apimachineryvalidation.NameIsDNS1035Label(workloadName, false)
+	if len(allErrs) != 0 {
+		g.logger.Info("couldn't generate Workload from kuma.io/workload label, contains invalid characters", "value", workloadName, "error", allErrs, "dataplane", dataplane.GetMeta().GetName())
+		return "", false
+	}
+	return workloadName, true
+}
+
+func (g *Generator) generate(ctx context.Context, mesh string, dataplanes []*core_mesh.DataplaneResource, workloads []*workload_api.WorkloadResource) {
+	log := g.logger.WithValues("mesh", mesh)
+
+	workloadsByName := g.collectWorkloadsByName(dataplanes)
+
+	for _, workload := range workloads {
+		workloadName := workload.GetMeta().GetName()
+		stillExists := workloadsByName[workloadName]
+
+		g.handleExistingWorkload(ctx, log, workload, stillExists)
```

Comment:
Func name `handleExistingWorkload` already implies `workload` exists, passing `stillExists` just to return is a bit odd.

I'd just inline the call to `handleExistingWorkload`.